### ISSUE TITLE
Regression bug 7.2.0 - Unable to cast DateTime into DataRow DateTime column

### ISSUE THREAD

**Issue by ili101** [PowerShell/PowerShell#17627]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```pwsh
$DataTable = [Data.DataTable]::new()
$null = $DataTable.Columns.Add('Date', [DateTime])
$Date = Get-Date
$Row = $DataTable.Rows.Add($Date)

$Row.Date = $Date # Error
$Row['Date'] = $Date # Error
$Row.Item('Date') = $Date # Works
$Row.Date = [DateTime]$Date # Works
```

Last working versions "PowerShell-7.2.0-preview.1" and "PowerShell-7.1.7".
First impacted version "PowerShell-7.2.0-preview.2".

### Expected behavior

`$Row.Date = $Date` will put the DateTime object in the DateTime column.

### Actual behavior

SetValueInvocationException: Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022 19:42:04> in Date Column.  Expected type is DateTime."

### Error details

```console
Exception             :
    Type           : System.Management.Automation.SetValueInvocationException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022
19:58:39> in Date Column.  Expected type is DateTime."
            HResult : -2146233087
        CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : CatchFromBaseAdapterSetValue
        InvocationInfo        :
            ScriptLineNumber : 5
            OffsetInLine     : 1
            HistoryId        : -1
            ScriptName       : C:\Users\illym\Desktop\Temp\PBug.ps1
            Line             : $Row.Date = $Date
            PositionMessage  : At C:\Users\illym\Desktop\Temp\PBug.ps1:5 char:1
                               + $Row.Date = $Date
                               + ~~~~~~~~~~~~~~~~~
            PSScriptRoot     : C:\Users\illym\Desktop\Temp
            PSCommandPath    : C:\Users\illym\Desktop\Temp\PBug.ps1
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, C:\Users\illym\Desktop\Temp\PBug.ps1: line 5
                                at <ScriptBlock>, <No file>: line 1
    TargetSite     :
        Name          : BasePropertySet
        DeclaringType : System.Management.Automation.Adapter, System.Management.Automation, Version=7.3.0.5, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message        : Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022
19:58:39> in Date Column.  Expected type is DateTime."
    Data           : System.Collections.ListDictionaryInternal
    InnerException :
        Type           : System.ArgumentException
        Message        : Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022 19:58:39> in Date Column.
Expected type is DateTime.
        TargetSite     :
            Name          : set_Item
            DeclaringType : System.Data.DataColumn
            MemberType    : Method
            Module        : System.Data.Common.dll
        InnerException :
            Type       : System.InvalidCastException
            TargetSite :
                Name          : ChkCast_Helper
                DeclaringType : System.Runtime.CompilerServices.CastHelpers, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
                MemberType    : Method
                Module        : System.Private.CoreLib.dll
            Message    : Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.
            Source     : System.Private.CoreLib
            HResult    : -2147467262
            StackTrace :
   at System.Data.Common.DateTimeStorage.Set(Int32 record, Object value)
   at System.Data.DataColumn.set_Item(Int32 record, Object value)
        Source         : System.Data.Common
        HResult        : -2147024809
        StackTrace     :
   at System.Data.DataColumn.set_Item(Int32 record, Object value)
   at System.Data.DataRow.set_Item(DataColumn column, Object value)
   at System.Management.Automation.DataRowAdapter.PropertySet(PSProperty property, Object setValue, Boolean convertIfPossible)
   at System.Management.Automation.Adapter.BasePropertySet(PSProperty property, Object setValue, Boolean convert)
    Source         : System.Management.Automation
    HResult        : -2146233087
    StackTrace     :
   at System.Management.Automation.Adapter.BasePropertySet(PSProperty property, Object setValue, Boolean convert)
   at System.Management.Automation.PSProperty.SetAdaptedValue(Object setValue, Boolean shouldConvert)
   at System.Management.Automation.PSProperty.set_Value(Object value)
   at System.Management.Automation.Language.PSSetMemberBinder.SetAdaptedValue(Object obj, String member, Object value)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : NotSpecified: (:) [], SetValueInvocationException
FullyQualifiedErrorId : CatchFromBaseAdapterSetValue
InvocationInfo        :
    ScriptLineNumber : 5
    OffsetInLine     : 1
    HistoryId        : -1
    ScriptName       : C:\Users\illym\Desktop\Temp\PBug.ps1
    Line             : $Row.Date = $Date
    PositionMessage  : At C:\Users\illym\Desktop\Temp\PBug.ps1:5 char:1
                       + $Row.Date = $Date
                       + ~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\Users\illym\Desktop\Temp
    PSCommandPath    : C:\Users\illym\Desktop\Temp\PBug.ps1
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, C:\Users\illym\Desktop\Temp\PBug.ps1: line 5
                        at <ScriptBlock>, <No file>: line 1
```

**Comment by fMichaleczek** [comment]
@ili101 So the problem is next to the cmdlet Get-Date

```
$DataTable = [Data.DataTable]::new()
$null = $DataTable.Columns.Add('Date', [DateTime])

$Date = [datetime]::Now
$Row = $DataTable.Rows.Add($Date)

$Row.Date = $Date # Works
$Row['Date'] = $Date # Works
$Row.Item('Date') = $Date # Works
$Row.Date = [DateTime]$Date # Works

$Row.Date = Get-Date # Error
$Row.Date = (Get-Date).psobject.ImmediateBaseObject # Works

```

**Comment by iSazonov** [comment]
I can not reproduce with latest build. Please download latest daily build and check with it.

**Comment by daxian-dbw** [comment]
The exception happens in `DateTimeStorage.Set(int record, object value)` when it cast the value to `IConvertable`: https://source.dot.net/#System.Data.Common/System/Data/Common/DateTimeStorage.cs,165

My investigation shows there has been no change in the related binder code and `DataRowAdapter` code. The version `7.2.0-preview.1` and `7.2.0-preview.2` were all on top of .NET 5.0, meaning they should all hit the `DateTimeStorage.Set` method. So, it must be a regression caused by [the changes in `7.2.0-preview.2`](https://github.com/PowerShell/PowerShell/releases/tag/v7.2.0-preview.2), but it's still not clear which one is the culprit. I will spend some more time on this the next Community Day.

**Comment by daxian-dbw** [comment]
It turns out the regression was caused by the change to the method `IsValueTypeMutable` in #12495, which was included in 7.2.0-preview.2. However, that was a correct change -- before the change, `IsValueTypeMutable` returns `true` for the type `DateTime` while it should have returned `false`, because `DateTime` is indeed immutable.

Before that change, due to `IsValueTypeMutable` returning `true`, `$Date = Get-Date` resulted in saving the base `DateTime` object to `$Date` (the `PSObject` wrapper was removed as PowerShell thought the base object is a mutable value type), and that's why `$Row.Date = $Date` and `$Row['Date'] = $Date` both worked fine.

After the change, `$Date = Get-Date` results in saving the PSObject-wrapped value to `$Date`, and that causes `$Row.Date = $Date` and `$Row['Date'] = $Date` to fail because internally, `$Row` expects a value of the `DateTime` type.

- The `$Row.Date = $Date` case is sort of easy to deal with, because as is mentioned by @fMichaleczek, it's our adapter code.
- The `$Row['Date'] = $Date` case is a little tricky, because it's a real set-indexer operation of the `System.Data.DataRow` type, with the signature to be `this[](string, object)` -- it takes an object. So, it's not clear that we should unwrap a PSObject here.

Need more thoughts on the right fix for this issue.

**Comment by daxian-dbw** [comment]
As proposed in https://github.com/PowerShell/PowerShell/pull/19071#discussion_r1092808924, it's probably better to have `PSObject` implements `System.IConvertible`, which will solve this problem and the other problems @ili101 mentioned in https://github.com/PowerShell/PowerShell/pull/19071#issuecomment-1413723688 without having to special-case any types in the binder code.
